### PR TITLE
9449 chainHub durability

### PR DESCRIPTION
**Pull request by turadg**
refs: #9449

## Description

This makes ChainHub return durable vows. The underlying lookup functions are not durable but they made so by the `retriable` helper.

So this also introduces a stub version of https://github.com/Agoric/agoric-sdk/issues/9541 pulled from @dtribble 's https://github.com/Agoric/agoric-sdk/pull/9657

### Security Considerations
none

### Scaling Considerations
none

### Documentation Considerations
none

### Testing Considerations
CI

### Upgrade Considerations
not yet upgradable. Communicated by https://github.com/Agoric/agoric-sdk/blob/a267ba20db0eeb2082c287ca7b69e37d6f272eb9/packages/vow/src/tools.js#L37-L38

### CODE REVIEW

File: packages/orchestration/src/exos/chain-hub.js [hunk 1]
```diff
@@ -133,19 +128,26 @@ export const makeChainHub = (agoricNames, zone = makeHeapZone()) => {
...
+          await null;
+          try {
+            const chainInfo = await E(agoricNames).lookup(CHAIN_KEY, chainName);
+            chainInfos.init(chainName, chainInfo);
+            return chainInfo;
+          } catch (_) {
+            throw makeError(`chain not found:${chainName}`);
+          }
         },
-      });
```
Comment (mhofman) [inline packages/orchestration/src/exos/chain-hub.js]: This needs to be defined not nested in a function call as it's a prepare pattern. Sorry that's a fidelity issue with the current stub

File: packages/orchestration/src/exos/chain-hub.js [hunk 2]
```diff
@@ -167,18 +169,28 @@ export const makeChainHub = (agoricNames, zone = makeHeapZone()) => {
...
+            return connectionInfo;
+          } catch (_) {
+            throw makeError(`connection not found: ${chainId1}<->${chainId2}`);
+          }
         },
-        onRejected: _cause => {
-          throw makeError(`connection not found: ${chainId1}<->${chainId2}`);
-        },
-      });
+      );
```
Comment (mhofman) [inline packages/orchestration/src/exos/chain-hub.js]: Same, must be defined top-level

File: packages/vow/src/tools.js [hunk 3]
```diff
@@ -33,6 +34,28 @@ export const prepareVowTools = (zone, powers = {}) => {
...
+   * @param {string} name
+   * @param {(...args: unknown[]) => ERef<T>} fn
+   * @returns {(...args: unknown[]) => Vow<T>}
+   */
+  const retriable =
+    (fnZone, name, fn) =>
+    (...args) => {
+      // TODO this needs to be an anonymous durable zone for the watched function
+      // Ideally, it should be a lazy one that only allocates the subzone if it gets used
+      const fnSubzone = makeHeapZone(`${name} hack subzone`);
```
Comment (mhofman) [inline packages/vow/src/tools.js]: Let's remove this for now. I am not convinced we need this in the final version

File: packages/vow/src/tools.js [hunk 4]
```diff
@@ -33,6 +34,28 @@ export const prepareVowTools = (zone, powers = {}) => {
...
+   * The internal functions
+   *
+   * @template T
+   * @param {Zone} fnZone - the zone for the named function
+   * @param {string} name
+   * @param {(...args: unknown[]) => ERef<T>} fn
+   * @returns {(...args: unknown[]) => Vow<T>}
+   */
+  const retriable =
+    (fnZone, name, fn) =>
```
Comment (mhofman) [inline packages/vow/src/tools.js:50]: Wondering if we should put a dummy `fnZone.prepareExo` here to force the user to call `retriable` top level, even though that would only let us detect issues during upgrades.

**Review by mhofman — CHANGES_REQUESTED** [review]

File: packages/orchestration/src/exos/chain-hub.js [hunk 5]
```diff
@@ -133,19 +153,16 @@ export const makeChainHub = (agoricNames, zone = makeHeapZone()) => {
...
-        },
-        onRejected: _cause => {
-          throw makeError(`chain not found:${chainName}`);
-        },
-      });
+      const lookup = vowTools.retriable(
+        zone,
+        'lookupChainInfo',
+        lookupChainInfo,
+      );
```
Comment (mhofman) [inline packages/orchestration/src/exos/chain-hub.js]: Sorry I wasn't clear earlier but the `vowTools.retriable` call needs to be done outside the method call, likely where `lookupChainInfo` is currently defined.

File: packages/orchestration/src/exos/chain-hub.js [hunk 6]
```diff
@@ -108,6 +103,31 @@ export const makeChainHub = (agoricNames, zone = makeHeapZone()) => {
     valueShape: IBCConnectionInfoShape,
   });
 
+  // eslint-disable-next-line no-restricted-syntax -- TODO more exact rules for vow best practices
+  const lookupChainInfo = async chainName => {
+    await null;
+    try {
+      const chainInfo = await E(agoricNames).lookup(CHAIN_KEY, chainName);
+      chainInfos.init(chainName, chainInfo);
```
Comment (mhofman) [inline packages/orchestration/src/exos/chain-hub.js]: Is there no risk of race with `registerChain` doing an `init` of the same name?

**Review by mhofman — CHANGES_REQUESTED** [review]

File: packages/orchestration/src/exos/chain-hub.js (hunk 5 above)
Comment (turadg) [reply packages/orchestration/src/exos/chain-hub.js]: ah that makes sense

File: packages/orchestration/src/exos/chain-hub.js (hunk 6 above)
Comment (turadg) [reply packages/orchestration/src/exos/chain-hub.js]: thanks

**Review by mhofman — APPROVED** [review]

**Pull request merged**